File: Leetcode/Top Interview Questions/289.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        # 0 -> 1: 2
        # 1 -> 0: 3

        dx = [-1, -1, -1, 0, 0, 1, 1, 1]
        dy = [-1, 0, 1, -1, 1, -1, 0, 1]

        n = len(board)
        m = len(board[0])
        for x in range(n):
            for y in range(m):
                count = 0
                for i in range(8):
                    nx = x + dx[i]
                    ny = y + dy[i]
                    if 0 <= nx < n and 0 <= ny < m and (board[nx][ny] == 1 or board[nx][ny] == 3):
                        count += 1
                if board[x][y] == 1:
                    if count < 2 or count > 3:
                        board[x][y] = 3
                else:
                    if count == 3:
                        board[x][y] = 2

        for x in range(n):
            for y in range(m):
                if board[x][y] == 2:
                    board[x][y] = 1
                elif board[x][y] == 3:
                    board[x][y] = 0

File: Leetcode/Top Interview Questions/289.js
/**
 * @param {number[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var gameOfLife = function(board) {
    const n = board.length
    const m = board[0].length
    
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            let liveCellCount = getLiveCellCount(board, i, j)

            if (board[i][j] === 1) {
                if (liveCellCount < 2 || liveCellCount > 3) board[i][j] = 3
            } else if (liveCellCount === 3) {
                board[i][j] = 2
            }
        }
    }

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (board[i][j] > 1) board[i][j] = 3 - board[i][j]
        }
    }
};

const getLiveCellCount = function(board, x, y) {
    const n = board.length
    const m = board[0].length
    const dx = [-1, -1, -1, 0, 0, 1, 1, 1]
    const dy = [-1, 0, 1, -1, 1, -1, 0, 1]
    let count = 0

    for (let i = 0; i < 8; i++) {
        let nx = x + dx[i]
        let ny = y + dy[i]
        if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
            if (board[nx][ny] === 1 || board[nx][ny] === 3) count++
        }
    } 

    return count
}
